File: examples/python/callback-json.py
import os
import json
import hmac
import hashlib
from collections import OrderedDict
from flask import Flask, request, jsonify

"""
iPaymu Callback Handler - Python/Flask (JSON)

Server berjalan di port 5001
Format: application/json

Cara menjalankan:
    pip install -r requirements.txt
    python callback-json.py
"""

app = Flask(__name__)

SECRET_KEY = os.getenv('SECRET_KEY', '123456')

def normalize_data(raw_data):
    result = {}
    for key, val in raw_data.items():
        if key == 'is_escrow':
            if str(val).lower() == 'true' or str(val) == '1' or val is True or val == 1:
                result[key] = 1
            else:
                result[key] = 0
        elif key in ['trx_id', 'status_code', 'transaction_status_code', 'paid_off', 'sub_total', 'total', 'amount', 'fee']:
            try:
                result[key] = int(val)
            except (ValueError, TypeError):
                result[key] = val
        elif key == 'additional_info':
            if val == '[]' or val == '':
                result[key] = []
            elif isinstance(val, str):
                try:
                    parsed = json.loads(val)
                    result[key] = parsed if isinstance(parsed, list) else []
                except json.JSONDecodeError:
                    result[key] = []
            elif isinstance(val, list):
                result[key] = val
            else:
                result[key] = []
        else:
            result[key] = val
            
    if 'additional_info' not in result:
        result['additional_info'] = []
        
    return result

def php_ksort(data):
    return OrderedDict(sorted(data.items(), key=lambda x: x[0]))

def generate_signature(data):
    normalized_data = normalize_data(data)
    normalized_data.pop('signature', None)
    sorted_data = php_ksort(normalized_data)
    
    json_body = json.dumps(sorted_data, separators=(',', ':'))
    json_body = json_body.replace('/', '\\/')
    
    signature = hmac.new(
        SECRET_KEY.encode('utf-8'),
        json_body.encode('utf-8'),
        hashlib.sha256
    ).hexdigest()
    
    return signature

@app.route('/callback', methods=['POST'])
def callback_handler():
    data = request.get_json()
    if not data:
        return jsonify({'status': 'error', 'message': 'Invalid JSON'}), 400
        
    received_signature = request.headers.get('X-Signature') or data.get('signature', '')
    
    if not received_signature:
        return jsonify({'status': 'error', 'message': 'Missing signature'}), 400
        
    calculated_signature = generate_signature(data)
    
    if hmac.compare_digest(calculated_signature, received_signature):
        print('✅ Signature valid!')
        return jsonify({'status': 'OK', 'message': 'Callback processed successfully'}), 200
    else:
        print(f'❌ Invalid signature! Expected: {received_signature}, Got: {calculated_signature}')
        return jsonify({
            'status': 'error',
            'message': 'Invalid Signature',
            'debug': {
                'received': received_signature,
                'calculated': calculated_signature
            }
        }), 400

@app.route('/health', methods=['GET'])
def health_check():
    return jsonify({
        'status': 'OK',
        'service': 'iPaymu Callback (JSON)'
    })

if __name__ == '__main__':
    port = int(os.getenv('PORT', 5001))
    print('═══════════════════════════════════════════════════')
    print('  iPaymu Callback Handler - JSON (Python)')
    print('═══════════════════════════════════════════════════')
    print(f'  Server running on http://localhost:{port}')
    print(f'  Endpoint: POST http://localhost:{port}/callback')
    print('═══════════════════════════════════════════════════')
    app.run(debug=True, host='0.0.0.0', port=port)

File: examples/javascript/callback-json.js
/**
 * iPaymu Callback Handler - JavaScript (JSON)
 * 
 * Server berjalan di port 3001
 * Format: application/json
 * 
 * Cara menjalankan:
 *   npm install
 *   npm run start:json
 * 
 * Atau langsung:
 *   node callback-json.js
 */

const crypto = require('crypto');
const express = require('express');

const app = express();

// Parse JSON bodies
app.use(express.json());

// Secret Key - Ganti dengan Nomor VA Anda!
const SECRET_KEY = process.env.SECRET_KEY || '123456';

/**
 * Normalisasi data dari JSON input
 * iPaymu mengirim JSON dengan tipe string, perlu dikonversi
 */
function normalizeData(rawData) {
  const result = {};

  for (const key in rawData) {
    let val = rawData[key];

    if (key === 'is_escrow') {
      // Integer: 0 atau 1
      result[key] = (val === 'true' || val === '1' || val === 1 || val === true) ? 1 : 0;
    }
    else if (['trx_id', 'status_code', 'transaction_status_code', 'paid_off', 'sub_total', 'total', 'amount', 'fee'].includes(key)) {
      // Integer
      result[key] = parseInt(val, 10);
    }
    else if (key === 'additional_info') {
      // Array: pastikan array
      if (Array.isArray(val)) {
        result[key] = val;
      } else if (val === '[]' || val === '') {
        result[key] = [];
      } else {
        try {
          result[key] = typeof val === 'string' ? JSON.parse(val) : val;
        } catch {
          result[key] = [];
        }
      }
    }
    else {
      // Tetap seperti aslinya (termasuk null)
      result[key] = val;
    }
  }

  // Pastikan additional_info ada
  if (!result.hasOwnProperty('additional_info')) {
    result['additional_info'] = [];
  }

  return result;
}

/**
 * Sort keys seperti PHP ksort (ascending A-Z, case-sensitive)
 */
function phpKsort(obj) {
  return Object.keys(obj)
    .sort((a, b) => a.localeCompare(b))
    .reduce((sortedObj, key) => {
      sortedObj[key] = obj[key];
      return sortedObj;
    }, {});
}

/**
 * Generate HMAC-SHA256 signature
 */
function generateSignature(data) {
  // Normalisasi data (iPaymu kirim string dalam JSON)
  const normalizedData = normalizeData(data);

  // Hapus signature jika ada
  delete normalizedData.signature;

  // Sort keys (A-Z)
  const sortedData = phpKsort(normalizedData);

  // Convert ke JSON dengan escape slash
  let jsonBody = JSON.stringify(sortedData);
  jsonBody = jsonBody.replace(/\//g, '\\/');

  // Generate HMAC-SHA256
  return crypto
    .createHmac('sha256', SECRET_KEY)
    .update(jsonBody)
    .digest('hex');
}

// Endpoint callback
app.post('/callback', (req, res) => {
  // Ambil signature dari header atau body
  const receivedSignature = req.headers['x-signature'] || req.body.signature;

  if (!receivedSignature) {
    return res.status(400).json({
      status: 'error',
      message: 'Missing signature'
    });
  }

  // Generate signature dari data yang diterima
  const calculatedSignature = generateSignature(req.body);

  // Validasi signature
  if (calculatedSignature === receivedSignature) {
    console.log('✅ Signature valid!');

    // TODO: Update status transaksi di database
    // TODO: Pastikan idempotency

    // Response sukses
    res.status(200).json({
      status: 'OK',
      message: 'Callback processed successfully'
    });
  } else {
    console.log(`❌ Invalid signature! Expected: ${receivedSignature}, Got: ${calculatedSignature}`);

    res.status(400).json({
      status: 'error',
      message: 'Invalid Signature'
    });
  }
});

const PORT = process.env.PORT || 3001;

app.listen(PORT, () => {
  console.log('═══════════════════════════════════════════════════');
  console.log('  iPaymu Callback Handler - JSON');
  console.log('═══════════════════════════════════════════════════');
  console.log(`  Server running on http://localhost:${PORT}`);
  console.log(`  Endpoint: POST http://localhost:${PORT}/callback`);
  console.log('═══════════════════════════════════════════════════');
});
